kusto/cloudinfo.go

```diff
@@ -0,0 +1,138 @@
+package kusto
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"strings"
+	"sync"
+)
+
+// abstraction to query metadata and use this information for providing all
+// information needed for connection string builder to provide all the requisite information
+
+const (
+	metadataEndpoint              = "v1/rest/auth/metadata"
+	defaultAuthEnvVarName         = "AadAuthorityUri"
+	defaultKustoClientAppId       = "db662dc1-0cfe-4e1c-a843-19a68e65be58"
+	defaultPublicLoginUrl         = "https://login.microsoftonline.com"
+	defaultRedirectUri            = "https://microsoft/kustoclient"
+	defaultKustoServiceResourceId = "https://kusto.kusto.windows.net"
+	defaultFirstPartyAuthorityUrl = "https://login.microsoftonline.com/f8cdef31-a31e-4b4a-93e4-5f571e91255a"
+	azureADKey                    = "AzureAD"
+)
+
+// retrieved metadata
+type cloudInfo struct {
+	loginEndpoint          string
+	loginMfaRequired       bool
+	kustoClientAppId       string
+	kustoClientRedirectUri string
+	kustoServiceResourceId string
+	firstPartyAuthorityUrl string
+}
+
+var defaultCloudInfo = &cloudInfo{
+	loginEndpoint:          getEnvOrDefault(defaultAuthEnvVarName, defaultPublicLoginUrl),
+	loginMfaRequired:       false,
+	kustoClientAppId:       defaultKustoClientAppId,
+	kustoClientRedirectUri: defaultRedirectUri,
+	kustoServiceResourceId: defaultKustoServiceResourceId,
+	firstPartyAuthorityUrl: defaultFirstPartyAuthorityUrl,
+}
+
+// cache to query it once per instance
+var cloudInfoCache map[string]*cloudInfo
+
+func RetrieveCloudInfoMetadata(kustoUrl string) (*cloudInfo, error) {
+	if cloudInfoCache != nil {
+		cachedCloudInfo, isExisting := cloudInfoCache[kustoUrl]
+		if isExisting {
+			return cachedCloudInfo, nil
+		}
+	} else {
+		//init the map here
+		cloudInfoCache = make(map[string]*cloudInfo)
+	}
+	var errorToReturn error
+
+	// there is no value for that URL that was picked.
+	var lock = &sync.Mutex{}
+	lock.Lock()
+	defer lock.Unlock()
+
+	// have to double check as the previous call may have succeeded and populated the value to the cache
+	if cloudInfoCache != nil {
+		cachedCloudInfo, isExisting := cloudInfoCache[kustoUrl]
+		if isExisting {
+			return cachedCloudInfo, nil
+		}
+	}
+
+	fullMetadataEndpoint := fmt.Sprintf("%s/%s", strings.TrimRight(kustoUrl, "/"), metadataEndpoint)
+	metadataResponse, err := http.Get(fullMetadataEndpoint)
+
+	var metadataMap map[string]map[string]interface{}
+
+	if err != nil {
+		// TODO how do we log
+		errorToReturn = err
+	}
+	// metadata retrieval was successful
+	if metadataResponse.StatusCode == 200 {
+		// close once read
+		defer metadataResponse.Body.Close()
+		jsonBytes, resError := ioutil.ReadAll(metadataResponse.Body)
```

Comment:
Let's make this way easiser:

```go
b, err := ioutil.ReadAll(resp.Body)
if err != nil {
  return CloudInfo{}, err // I'm assuming we do away with the pointer, so now we return the zero value
}

// This piece of code is suspect to me.  If you don't get a body, isn't that a server error.  And if it is
// and the result is always to just return defaultCloudInfo, then it seems we should do this differently.
// But I'm going to leave it in for the moment.
if len(b) == 0 {
  cloudInfoCache[kustoURL] = defaultCloudInfo
  return defaultCloudInfo
}

m := map[string]map[string]interface{}{} // This should really be a concrete type.
if err := json.Unmarshal(b, &m); err != nil { // You were ignoring the error here, which you shouldn't do.
  return CloudInfo{}, fmt.Errorf("metadata service returned bad JSON: %w", err)
}
...
```

I think you should have a concrete type I did not include above for this whole nested data, then you can just JSON marshal directly in.  This only requires one struct and making the CloudInfo fields public.

```go
type metaResp struct {
  AzureAD AzureADResp
}

type CloudInfo struct {
  LoginMFARequired bool `json: "LoginMfaRequired"`
  KustoClientAppID string `json: "KustoClientAppId"`
  KustoClientRedirectURI string `json: "KustoClientRedirectUri"`
  KustoServiceResourceID string `json: "KustoServiceResourceId"`
  FirstPartyAuthorityURL string `json: "FirstPartyAuthorityUrl"
}
```

Now you could just do this directly with:

```go 
md := metaResp{}
if err := json.Unmarshal(b, &md); err != nil {
  ...
}
```